### DISCUSSION TITLE
Aggregated habits widget

### DISCUSSION THREAD

**Discussion by r1cc4rdo** [iSoron/uhabits#1270] [Technical Discussion]
I'm a developer willing to contribute, but new to both kotlin and Android development.

A recurrent theme in the discussion forum is a feature request for a single widget with checkmarks for multiple habits, for example:
* #45
* #51
* #531
* #1073

Stacked widgets do not really answer these requests. As a person who struggles with motivation, having information hidden behind multiple swipes is a no-go.

Replies from contributors there have been along the lines of:
> "Such a widget is unlikely to be technically feasible (...) Android widgets are very restricted and it would be exceedingly hard to recreate the main app functionality in a widget. Even if we somehow managed to do it, it would likely be sluggish and have poor usability"

My understanding of the above is that the difficulty lies in replicating the app UI in the widget, which amounts to re-implementing a brittle version of the whole main app in a constrained environment.

How feasible would be though to just *display* a list of habits and checkmarks, in a widget that just opens the main app when tapped? This would cover all my user needs.

**Comment by iSoron** [comment]
Welcome to the project, @r1cc4rdo.

> How feasible would be though to just display a list of habits and checkmarks, in a widget that just opens the main app when tapped?

It may be feasible, although hard to implement. If you would like to try to implement a prototype, that would be welcome.

**Reply by r1cc4rdo** [reply]
Thank you for your reply, @iSoron.

> It may be feasible, although hard to implement. If you would like to try to implement a prototype, that would be welcome.

I'm willing to give it an honest try, but before that would you mind shedding a few more details?

It's honestly surprising that you deem the above hard: superficially, it looks like there no other thing to do than display information and open the main app, which current widgets do already. I'd rather know beforehand what I'm getting myself into :)

**Reply by iSoron** [reply]
@r1cc4rdo The main difficulty is that Android widgets are not allowed to have custom views. To get around this, we currently render custom views in-memory and display them as images in widgets, but if you try the same approach here, you would have issues with scrolling.

**Reply by r1cc4rdo** [reply]
I did play a bit with the code, and below is my underwhelming attempt.
![Screen Shot 2022-01-30 at 20 42 17](https://user-images.githubusercontent.com/12174207/151740097-457b2e44-dbbd-4a09-95f7-36790d294a34.png)

I created a novel widget, provider and chart, passing them a list of habits. I draw as many as they fit on the widget drawable area.

If there are more widget to show than the available space the idea, unimplemented, would be to overflow them in pages as StackedWidgets do.

Given the above, it seems to me that a multiple habit widget is doable?

At this point, I'm already very much out of my depth re: the codebase, Kotlin and Android development altogether. Heck, I decided to report here after trying four different ways to change the widget background color, and failing still :( There is likely no point in putting together a PR for unfinished ugly test code, but can do that if someone else is interested in taking over.

**Reply by iSoron** [reply]
@r1cc4rdo Thanks for doing some work on this. Even though it's unfinished, I think it would be nice to make the code available somewhere. It could serve as a starting point for someone else in the future.

Your idea of rendering the entire list as a "chart" might work, and it might be easier to implement. I was originally considering building the widget using multiple views (i.e. one for each habit). Your pagination idea might also work, although it might be hard to implement. Widgets seem to support a primitive form of scrolling ([through a ListView hack](https://stackoverflow.com/a/32646205)) so perhaps we could try that first.

**Reply by r1cc4rdo** [reply]
Thanks. I'll clean up and put it as a forked repo then.

**Reply by r1cc4rdo** [reply]
I've cleaned up the above and put it here: [r1cc4rdo/uhabits](https://github.com/r1cc4rdo/uhabits).
